src/kinsol/kinsol.c

```diff
@@ -2676,8 +2677,30 @@ static int KINPicardAA(KINMem kin_mem)
     /* compute new solution */
     if (kin_mem->kin_m_aa == 0 || kin_mem->kin_nni - 1 < kin_mem->kin_delay_aa)
     {
-      if (kin_mem->kin_damping)
+      if (kin_mem->kin_damping || kin_mem->kin_damping_fn)
       {
+        if (kin_mem->kin_damping_fn)
+        {
+          retval = kin_mem->kin_damping_fn(kin_mem->kin_nni, kin_mem->kin_uu,
+                                           kin_mem->kin_fval, 0, SUN_RCONST(-1.0),
+                                           kin_mem->kin_user_data,
+                                           &(kin_mem->kin_beta));
+          if (retval)
+          {
+            KINProcessError(kin_mem, KIN_DAMPING_FN_ERR, __LINE__, __func__,
+                            __FILE__, "The damping function failed.");
+            ret = KIN_DAMPING_FN_ERR;
+            break;
+          }
+          if (kin_mem->kin_beta <= ZERO)
+          {
+            KINProcessError(kin_mem, KIN_ILL_INPUT, __LINE__, __func__, __FILE__,
+                            "The damping parameter is negative or zero.");
+            ret = KIN_ILL_INPUT;
```

Comment:
Is this really an illegal input, as opposed to a damping function error?  I realize that the message is more important than the value of the error flag, but I'd think that using `KIN_DAMPING_FN_ERR` would make more sense.

If you choose to change it here, then the same change should be made a few additional times below